### PR TITLE
client,cmd/snap: first pass at client messaging around modes

### PR DESCRIPTION
**Pull request by chipaca**
more a strawman than anything else at this stage

### CODE REVIEW

File: cmd/snap/cmd_snap_op.go [hunk 1]
```diff
@@ -414,10 +414,15 @@ func (x *cmdInstall) installOne(name string, opts *client.SnapOptions) error {
...
 		changeID, err = cli.Install(name, opts)
 	}
-	if e, ok := err.(*client.Error); ok && e.Kind == client.ErrorKindSnapAlreadyInstalled {
-		fmt.Fprintf(Stderr, i18n.G("snap %q is already installed, see \"snap refresh --help\"\n"), name)
+	if e, ok := err.(*client.Error); ok {
+		msg, isError := clientErrorToCmdMessage(name, e)
+		if isError {
+			return errors.New(msg)
+		}
+		fmt.Fprintf(Stderr, msg)
```
Comment (zyga) [inline cmd/snap/cmd_snap_op.go]: If it's not an error then maybe we want to print this to stdout?

File: cmd/snap/error.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	case client.ErrorKindSnapAlreadyInstalled:
+		return fmt.Sprintf(i18n.G(`snap %q is already installed, see "snap refresh --help"`), snapName), false
+	case client.ErrorKindSnapNeedsMode:
+		switch err.Value {
+		case client.DevModeConfinement:
+			msg = i18n.G(`
+the developer of snap %q has indicated that they do not consider it to
+be of production quality, and is only meant for developers or user
+testing at this point.  Installing this snap from an untrusted developer
+can put your system at risk.  What's more this developer-mode snap once
```
Comment (zyga) [inline cmd/snap/error.go]: I'd add a comma after `What's more`

File: cmd/snap/error.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+installed will need to be manually refreshed.  If all of the above is
+agreeable to you please repeat this command with the --devmode option
+added.
+`)
+
+		case client.ClassicConfinement:
+			msg = i18n.G(`
+snap %q is a "classic" snap, which means it runs outside of the sandbox
+more strict snaps run in.  As this could put your system at risk (like
```
Comment (zyga) [inline cmd/snap/error.go]: I'd reword this to 
```
snap %q uses "classic" confinement, which means that it is not confined at all and has full
access to your system (like a typical classic package on a Linux distribution). ...
```

File: cmd/snap/error.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	"bufio"
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+)
+
+func fill(para string) string {
```
Comment (zyga) [inline cmd/snap/error.go]: I tried to parse this and figure out what is happening exactly but I bailed out. Can you add a comment please?

File: client/client.go [hunk 5]
```diff
@@ -318,8 +318,9 @@ type response struct {
 
 // Error is the real value of response.Result when an error occurs.
 type Error struct {
-	Kind    string `json:"kind"`
-	Message string `json:"message"`
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
```
Comment (pedronis) [inline client/client.go]: wasn't it more natural to have no Value and have two kinds instead ?

File: cmd/snap/error.go (hunk 4 above)
Comment (pedronis) [inline cmd/snap/error.go]: why fill? isn't this doing wrap-around?

File: cmd/snap/cmd_snap_op.go (hunk 1 above)
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go]: I think stderr is fine

File: cmd/snap/error.go (hunk 3 above)
Comment (pedronis) [inline cmd/snap/error.go]: where does the phrasing come from?   --classic is nuanced because on one hand it indeed gives access to all these things, otoh at least from the store is enabled only for "vetted" publishers

File: cmd/snap/error.go (hunk 4 above)
Comment (chipaca) [inline cmd/snap/error.go]: Because that's what `textwrap` calls something like this; `fill` there returns a string, whereas `wrap` returns a list of lines.

I'm not saying it's a particularly *good* reason, but that's why I called it that.

File: cmd/snap/error.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+func clientErrorToCmdMessage(snapName string, err *client.Error) (msg string, isError bool) {
+	switch err.Kind {
+	case client.ErrorKindSnapAlreadyInstalled:
+		return fmt.Sprintf(i18n.G(`snap %q is already installed, see "snap refresh --help"`), snapName), false
+	case client.ErrorKindSnapNeedsMode:
+		switch err.Value {
+		case client.DevModeConfinement:
+			msg = i18n.G(`
+the developer of snap %q has indicated that they do not consider it to
```
Comment (chipaca) [inline cmd/snap/error.go]: Also: we should mention `--jailmode` here.

Comment (chipaca) [inline cmd/snap/error.go]: e.g. change the last sentence about devmode to `f you understand and want to proceed repeat the command including --devmode; if instead you want to install the snap forcing it into strict confinement repeat the command including --jailmode.`

**Comment by niemeyer** [comment]
@chipaca Easy win here. :)

**Comment by zyga** [comment]
Some test failures look genuine:
```
FAIL: main_test.go:185: SnapSuite.TestAccessDeniedHint

main_test.go:195:
    c.Check(err.Error(), Equals, `access denied (try with sudo)`)
... obtained string = "access denied"
... expected string = "access denied (try with sudo)"
```

**Comment by chipaca** [comment]
yeap, those are the ones i mentioned in the standup a few days ago

File: cmd/snap/cmd_snap_op.go [hunk 7]
```diff
@@ -414,10 +414,15 @@ func (x *cmdInstall) installOne(name string, opts *client.SnapOptions) error {
 	} else {
 		changeID, err = cli.Install(name, opts)
 	}
-	if e, ok := err.(*client.Error); ok && e.Kind == client.ErrorKindSnapAlreadyInstalled {
-		fmt.Fprintf(Stderr, i18n.G("snap %q is already installed, see \"snap refresh --help\"\n"), name)
+	if e, ok := err.(*client.Error); ok {
+		msg, isError := clientErrorToCmdMessage(name, e)
```
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: I struggled to understand your point here, as nothing seemed to fit a "usual pattern", but suddenly it hit me that maybe what you're talking about is just about returning error as the second argument instead of a boolean. If that's clearer, sure, I'll do that!

Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Yes, the usual pattern for handling errors as a whole: how it's returned, how it's tested, how it's passed through.

File: client/client.go (hunk 5 above)
Comment (niemeyer) [inline client/client.go]: Still an open point it seems.

Comment (niemeyer) [inline client/client.go]: Is the field a left over now?

File: cmd/snap/cmd_snap_op.go (hunk 7 above)
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Still pending as well, or at least pending a rejection. :)

File: cmd/snap/cmd_snap_op.go (hunk 1 above)
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Yes, stderr is the right place to send warnings towards human eyes instead of the next tool in the pipeline. Strictly speaking, both cases are an error, actually. It's just that in one of them we have a text to present to the user instead of it being just a traditional error line.

File: client/client.go (hunk 5 above)
Comment (chipaca) [inline client/client.go]: I left it because it's documented, the server exposes it, this was just the first time it was getting used client-side

File: cmd/snap/cmd_snap_op.go (hunk 7 above)
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: I've done this refactor, although it doesn't seem cleaner nor clearer to me now but I'm the one writing it so :-)

File: cmd/snap/error.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	// 3 is the %v\n, which will be present in any locale
+	indent := len(errorPrefix) - 3
+	var buf bytes.Buffer
+	doc.ToText(&buf, para, strings.Repeat(" ", indent), "", width)
+
+	return strings.TrimSpace(buf.String())
+}
+
+func clientErrorToCmdMessage(snapName string, err *client.Error) (string, error) {
+	isError := true
```
Comment (pedronis) [inline cmd/snap/error.go]: maybe also have a mentionsSnap flag? and avoid the returns in the middle of the switch?  I found them a bit confusing but maybe it's just me

File: cmd/snap/error.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	if f, ok := Stdout.(*os.File); ok {
+		width, height, _ = terminal.GetSize(int(f.Fd()))
+	}
+
+	if width <= 0 {
+		width = int(osutil.GetenvInt64("COLUMNS"))
+	}
+
+	if height <= 0 {
+		height = int(osutil.GetenvInt64("COLUMNS"))
```
Comment (chipaca) [inline cmd/snap/error.go]: /o\!!!
every time i read past that something seemed off. Thank you!
